tanglegram/tangle.py

```diff
@@ -267,137 +207,110 @@ def get_all_linkage_gen(linkage, stop, labels, start=0):
-    org_entang = float(min_entang)
-
-    lindex1 = leaf_order(best_linkage1, labels1, as_dict=True)
-    lindex2 = leaf_order(best_linkage2, labels2, as_dict=True)
-
-    # For each label
-    for k in list(lindex1):
-        find1 = lindex1[k]
-        find2 = lindex2[k]
-        # If this label is not aligned between left and right dendrogram
-        if find1 != find2:
-            # Find the first hinges for this label
-            knot1 = np.where(best_linkage1 == find1)[0][0]
-            knot2 = np.where(best_linkage2 == find2)[0][0]
-
-            # Rotate around these hinges
+def refine(best_linkage1, best_linkage2, labels1, labels2):
+    """Rotate dendrogram by going through all objects to find smallest entanglement."""
+    dend1 = dendrogram(best_linkage1, labels=labels1, no_plot=True)
+    dend2 = dendrogram(best_linkage2, labels=labels2, no_plot=True)
+    old_entang = entanglement(best_linkage1, best_linkage2)
+    min_entang = old_entang
+
+    # If an object does not appear at the same position on x-axis in both dendrograms
+    # Find the first interior vertex containing such object and rotate dendrograms
+    for i in range(len(best_linkage1)):
+        if dend1["ivl"][i] != dend2["ivl"][i]:
+            for num, item in enumerate(best_linkage1):
+                if item[0] == i or item[1] == i:
+                    knot1 = num
+            for num, item in enumerate(best_linkage2):
+                if item[0] == i or item[1] == i:
+                    knot2 = num
             new1 = rotate(best_linkage1, knot1)
             new2 = rotate(best_linkage2, knot2)
             all1 = np.append([best_linkage1], [new1], axis=0)
             all2 = np.append([best_linkage2], [new2], axis=0)
-
-            all1_lindices = []
+            save = []
             for j in all1:
-                all1_lindices.append(leaf_order(j, labels1, as_dict=True))
-
-            all2_lindices = []
-            for k in all2:
-                all2_lindices.append(leaf_order(k, labels2, as_dict=True))
-
-            # Check if any of the new versions are better than the old
-            for j, lix1 in zip(all1, all1_lindices):
-                for k, lix2 in zip(all2, all2_lindices):
-                    new_entang = entanglement(lix1, lix2, L=L)
+                for k in all2:
+                    new_entang = entanglement(j, k)
                     if new_entang < min_entang:
                         min_entang = new_entang
                         best_linkage1 = j
                         best_linkage2 = k
-                        lindex1 = lix1
-                        lindex2 = lix2
-
-    improved = min_entang < org_entang
+            save.append(min_entang)
+    improved = old_entang - min_entang
     return best_linkage1, best_linkage2, min_entang, improved
 
 
-def untangle(link1, link2, labels1, labels2, method='random', L=1.5, **kwargs):
+def untangle(link1, link2, labels1, labels2, method='random', L=2.0, **kwargs):
```

Comment:
Looks like `labels` are still accepted but essentially ignored in favour of just using the linkage. This implies that the labels in each linkage always match perfectly (i.e. index left 1 = index 1 right and so on). This may work for toy examples but will not be true for most real world examples. This change is unfortunately a deal breaker.

Reply:
It is assumed that the set of objects (labels) in two dendrograms have a one-to-one correspondence. Such case occasionally occurs in real life when we apply different hierarchical clustering algorithms on the same dataset.